### PR TITLE
LibWeb: Port URL and URLSearchParams to new String

### PR DESCRIPTION
**Pull request by kennethmyhra**
Follwing changes to BindingsGenerator:
- LibWeb: Add new String support for parameters with empty string defaults
- LibWeb: Handle optional return values for getters returning new String

### CODE REVIEW

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp [hunk 1]
```diff
@@ -1487,11 +1487,19 @@ static void generate_wrap_statement(SourceGenerator& generator, DeprecatedString
...
+            if (!interface.extended_attributes.contains("UseNewAKString")) {
+                scoped_generator.append(R"~~~(
     if (@value@.is_null()) {
         @result_expression@ JS::js_null();
     } else {
 )~~~");
+            } else {
+                scoped_generator.append(R"~~~(
+    if (@value@.is_empty()) {
+        @result_expression@ JS::js_null();
```
Comment (linusg) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:1499]: That looks extremly wrong - these functions should return an `Optional<String>`, as a replacement for the `DeprecatedString` null state.

File: AK/URL.h [hunk 2]
```diff
@@ -47,6 +48,10 @@ class URL {
         : URL(string.view())
     {
     }
+    URL(String const& string)
+        : URL(string.bytes_as_string_view())
+    {
+    }
```
Comment (linusg) [inline AK/URL.h:54]: Why? `String` should be able to implicitly convert to `StringView` and invoke the appropriate constructor above

https://github.com/SerenityOS/serenity/blob/8f1d13e73b8bae4f86e5cf2ed7c42d757a6ac0a7/AK/StringView.h#L51

Comment (kennethmyhra) [reply AK/URL.h:54]: Without it the compiler complains about the conversions listed below.

https://github.com/kennethmyhra/serenity/blob/a658386b38735d7b834b4516456c48978f9e4879/Userland/Libraries/LibWeb/URL/URL.cpp#L30

https://github.com/kennethmyhra/serenity/blob/a658386b38735d7b834b4516456c48978f9e4879/Userland/Libraries/LibWeb/URL/URL.cpp#L40

https://github.com/kennethmyhra/serenity/blob/a658386b38735d7b834b4516456c48978f9e4879/Userland/Libraries/LibWeb/URL/URL.cpp#L101

I can get around it by doing this:
```cpp
// 1. Let parsedBase be the result of running the basic URL parser on base.
parsed_base = AK::URL { base.value() };
```
or this:
```cpp
// 1. Let parsedBase be the result of running the basic URL parser on base.
parsed_base = base.value().bytes_as_string_view();
```

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp (hunk 1 above)
Comment (kennethmyhra) [reply Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp]: Absolutely, should have known better. Fixed

File: AK/URL.h (hunk 2 above)
Comment (linusg) [reply AK/URL.h:54]: I see - not a huge fan of implicit URL parsing anyway, but let's preserve the status quo for now 👍

File: Userland/Libraries/LibWeb/URL/URLSearchParams.cpp [hunk 3]
```diff
@@ -156,39 +160,49 @@ size_t URLSearchParams::size() const
...
+    TRY_OR_THROW_OOM(vm, m_list.try_empend(name, value));
     // 2. Update this.
-    update();
+    TRY(update());
+
+    return {};
 }
 
-void URLSearchParams::update()
+JS::ThrowCompletionOr<void> URLSearchParams::update()
```
Comment (linusg) [inline Userland/Libraries/LibWeb/URL/URLSearchParams.cpp:175]: Hmm, why do we need a `ThrowCompletionOr` here and can't just propagate the `ExceptionOr` from `to_string()`?

File: Userland/Libraries/LibWeb/URL/URLSearchParams.cpp [hunk 4]
```diff
@@ -200,18 +214,18 @@ DeprecatedString URLSearchParams::get(DeprecatedString const& name)
 }
 
 // https://url.spec.whatwg.org/#dom-urlsearchparams-getall
-Vector<DeprecatedString> URLSearchParams::get_all(DeprecatedString const& name)
+Vector<String> URLSearchParams::get_all(String const& name)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/URL/URLSearchParams.cpp:217]: Could update this to use fallible vector append as well while we're here :)

**Review by linusg — COMMENTED** [review]
Looks great! one more question :)

File: Userland/Libraries/LibWeb/URL/URLSearchParams.cpp (hunk 3 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/URL/URLSearchParams.cpp]: Uh, I am not really sure, maybe trying to be clever? Anyway changed it to `WebIDL::ExceptionOr<>`

File: Userland/Libraries/LibWeb/URL/URLSearchParams.cpp (hunk 4 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/URL/URLSearchParams.cpp]: Done:)

**Pull request merged**